File: scripts/today_events.py
"""今日事(纯文本)—— 每天 17:00 北京自动出,落 ~/Downloads,复制即发。

自家源:Finnhub 经济+财报日历 + 自家 us-stocks 补公司名/市值。
比对手版多:预期/前值、市值上下文、重要度标记、自动"今日主角"。
不装假精度:财报只标 盘前/盘后(Finnhub 免费档没有精确到分的时间)。

用法: uv run python scripts/today_events.py [--date 2026-06-10]
"""
from __future__ import annotations
import argparse
import json
import os
import re
import sys
from datetime import datetime, timezone, timedelta
from pathlib import Path

import requests

ROOT = Path(__file__).parent.parent
try:
    from dotenv import load_dotenv; load_dotenv(ROOT / ".env")
except Exception:
    pass
FINN = os.environ.get("FINNHUB_KEY") or ""
OUT_DIR = Path.home() / "Downloads"
ET = timezone(timedelta(hours=-4))

# 宏观事件中文名(常见高频;没映射的保留英文)
ZH = [
    (r"Core Inflation Rate MoM", "核心CPI 环比"),
    (r"Core Inflation Rate YoY", "核心CPI 同比"),
    (r"Inflation Rate MoM", "CPI 环比"),
    (r"Inflation Rate YoY", "CPI 同比"),
    (r"Core PPI|PPI MoM|Producer Price.*MoM", "PPI 环比"),
    (r"PPI YoY|Producer Price.*YoY", "PPI 同比"),
    (r"Core PCE.*MoM", "核心PCE 环比"),
    (r"Fed Interest Rate Decision|FOMC", "美联储利率决议"),
    (r"Non.?Farm Payrolls", "非农就业"),
    (r"Unemployment Rate", "失业率"),
    (r"Initial Jobless Claims", "首申失业金"),
    (r"Retail Sales MoM", "零售销售 环比"),
    (r"GDP Growth Rate", "GDP"),
    (r"Michigan Consumer Sentiment", "密歇根消费者信心"),
    (r"ISM Manufacturing", "ISM 制造业"),
    (r"ISM Services", "ISM 服务业"),
    (r"MBA Mortgage Applications", "MBA 按揭申请"),
    (r"EIA Crude Oil Stocks Change", "EIA 原油库存"),
    (r"Monthly Budget Statement", "月度财政收支"),
    (r"(\d+)-Year Note Auction", r"\1年期国债拍卖"),
    (r"(\d+)-Year Bond Auction", r"\1年期长债拍卖"),
]
# 低重要度白名单(low 也保留的)
LOW_KEEP = re.compile(r"MBA Mortgage Applications|10-Year Note Auction|30-Year Bond Auction", re.I)
# 中重要度黑名单(medium 也丢的:同族噪音,保代表行即可)
MED_DROP = re.compile(r"MBA (30-Year|Mortgage Market|Mortgage Refinance|Purchase)|"
                      r"EIA (?!Crude Oil Stocks Change)|CPI s\.a|^CPI$", re.I)


def zh_name(en: str) -> str:
    for pat, rep in ZH:
        m = re.search(pat, en, re.I)
        if m:
            return re.sub(pat, rep, en, flags=re.I) if "\\1" in rep else rep
    return en


def fnum(x) -> str:
    if x in (None, ""):
        return ""
    try:
        v = float(x)
        return f"{v:g}"
    except (TypeError, ValueError):
        return str(x)


def macro_lines(date: str) -> list[str]:
    j = requests.get("https://finnhub.io/api/v1/calendar/economic", params={"token": FINN}, timeout=25).json()
    rows = []
    for e in j.get("economicCalendar") or []:
        if e.get("country") not in ("US", "United States"):
            continue
        try:
            dt = datetime.strptime(str(e.get("time", "")), "%Y-%m-%d %H:%M:%S").replace(tzinfo=timezone.utc).astimezone(ET)
        except ValueError:
            continue
        if dt.strftime("%Y-%m-%d") != date:
            continue
        name, imp = str(e.get("event", "")), str(e.get("impact", ""))
        if imp == "low" and not LOW_KEEP.search(name):
            continue
        if imp == "medium" and MED_DROP.search(name):
            continue
        if imp == "high" and re.search(r"CPI s\.a|^CPI$", name):
            continue
        mark = "🔴 " if imp == "high" else ""
        est, prev = fnum(e.get("estimate")), fnum(e.get("prev"))
        tail = " · ".join(x for x in [f"预期 {est}" if est else "", f"前值 {prev}" if prev else ""] if x)
        rows.append((dt.strftime("%H:%M"), f"{dt.strftime('%H:%M')}  {mark}{zh_name(name)}" + (f"  {tail}" if tail else "")))
    rows.sort()
    # 同名去重(Finnhub 偶有重复行)
    seen, out = set(), []
    for _, line in rows:
        key = re.sub(r"\s+", " ", line)
        if key not in seen:
            seen.add(key)
            out.append(line)
    return out


def earnings_lines(date: str) -> tuple[list[str], list[str], str]:
    us = {}
    try:
        us = {s["sym"]: (s.get("name"), s.get("mcapB")) for s in
              json.loads((ROOT / "web/public/data/us-stocks.json").read_text(encoding="utf-8"))["stocks"]}
    except Exception:
        pass
    j = requests.get("https://finnhub.io/api/v1/calendar/earnings",
                     params={"from": date, "to": date, "token": FINN}, timeout=25).json()
    ears = j.get("earningsCalendar") or []
    star = ("", 0.0)  # 主角财报(最大市值)

    def fmt(e) -> str | None:
        nonlocal star
        sym = e.get("symbol") or ""
        nm, mc = us.get(sym, (None, None))
        if mc and mc > star[1]:
            star = (sym, mc)
        eps = e.get("epsEstimate")
        bits = [f"${sym}"]
        if nm:
            bits.append(nm[:26])
        if mc and mc >= 1:
            bits.append(f"${mc:.0f}B")
        if isinstance(eps, (int, float)):
            bits.append(f"EPS est {eps:.2f}")
        return "  ".join(bits)

    def grp(hours: tuple[str, ...], cap: int) -> list[str]:
        g = [e for e in ears if (e.get("hour") or "") in hours]
        g.sort(key=lambda e: -(us.get(e.get("symbol"), (None, 0))[1] or 0))
        out, rest = [], []
        for e in g:
            mc = us.get(e.get("symbol"), (None, 0))[1] or 0
            (out if (len(out) < cap and mc >= 0.5) else rest).append(e)
        lines = [fmt(e) for e in out]
        if rest:
            small = [f"${e.get('symbol')}" for e in rest if e.get("symbol")][:8]
            if small:
                lines.append("小盘同日: " + " ".join(small))
        return lines

    return grp(("bmo",), 6), grp(("amc", "", "dmh"), 7), (f"${star[0]}" if star[0] else "")


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--date", default="")
    args = ap.parse_args()
    if not FINN:
        sys.exit("缺 FINNHUB_KEY")
    now = datetime.now(ET)
    date = args.date or now.strftime("%Y-%m-%d")
    wd = ["周一", "周二", "周三", "周四", "周五", "周六", "周日"][datetime.strptime(date, "%Y-%m-%d").weekday()]

    macro = macro_lines(date)
    bmo, amc, star = earnings_lines(date)
    hi_macro = [l for l in macro if "🔴" in l]
    lead_bits = []
    if hi_macro:
        m = re.search(r"(\d{2}:\d{2})\s+🔴\s+(\S+)", hi_macro[0])
        if m:
            lead_bits.append(f"{m.group(2).split(' ')[0]} {m.group(1)}")
    if star:
        lead_bits.append(f"{star} 盘后")

    L = [f"今日事 · {date.replace('-', '/')}({wd},全美东时间)", ""]
    if macro:
        L += ["【宏观】"] + macro + [""]
    if bmo:
        L += ["【盘前财报】"] + bmo + [""]
    if amc:
        L += ["【盘后财报】"] + amc + [""]
    if lead_bits:
        L += [f"今日主角:{' · '.join(lead_bits)}"]
    text = "\n".join(L).rstrip() + "\n"

    out = OUT_DIR / f"TodayEvents_{date.replace('-', '')}.txt"
    out.write_text(text, encoding="utf-8")
    print(text)
    print(f"→ {out}")


if __name__ == "__main__":
    main()

File: web/src/lib/today-events.ts
// 今日大事(盘报 tab 第二段)—— 服务端直连 Finnhub,30 分钟缓存,无需定时任务/部署。
// 口径:宏观降噪(MBA/EIA 系折叠、low 过滤带白名单、中文名映射、🔴=high)+
//       财报盘前/盘后分组(us-stocks 补名称/市值,≥$0.5B 上正文,小盘归一行)+ 自动主角。
import "server-only";
import { promises as fs } from "fs";
import path from "path";

export type MacroEvent = { t: string; hi: boolean; name: string; est: string; prev: string };
export type EarnRow = { sym: string; name: string; mcapB: number | null; eps: number | null };
export type IpoRow = { sym: string; name: string; exch: string; price: string; valB: number; date: string; isToday: boolean };
export type TodayEvents = {
  date: string;          // YYYY-MM-DD(ET)
  weekday: string;       // 周三
  macro: MacroEvent[];
  bmo: EarnRow[];
  amc: EarnRow[];
  smallBmo: string[];
  smallAmc: string[];
  ipos: IpoRow[];        // 当天 ≥$1B + 7天内 ≥$10B 巨无霸(提前展示)
  lead: string;          // "CPI 08:30 · $ORCL 盘后 · SpaceX IPO 周五"
};

const ZH: [RegExp, string][] = [
  [/Core Inflation Rate MoM/i, "核心CPI 环比"],
  [/Core Inflation Rate YoY/i, "核心CPI 同比"],
  [/Inflation Rate MoM/i, "CPI 环比"],
  [/Inflation Rate YoY/i, "CPI 同比"],
  [/Core PPI|PPI MoM|Producer Price.*MoM/i, "PPI 环比"],
  [/PPI YoY|Producer Price.*YoY/i, "PPI 同比"],
  [/Core PCE.*MoM/i, "核心PCE 环比"],
  [/Fed Interest Rate Decision|FOMC/i, "美联储利率决议"],
  [/Non.?Farm Payrolls/i, "非农就业"],
  [/Unemployment Rate/i, "失业率"],
  [/Initial Jobless Claims/i, "首申失业金"],
  [/Retail Sales MoM/i, "零售销售 环比"],
  [/GDP Growth Rate/i, "GDP"],
  [/Michigan Consumer Sentiment/i, "密歇根消费者信心"],
  [/ISM Manufacturing/i, "ISM 制造业"],
  [/ISM Services/i, "ISM 服务业"],
  [/MBA Mortgage Applications/i, "MBA 按揭申请"],
  [/EIA Crude Oil Stocks Change/i, "EIA 原油库存"],
  [/Monthly Budget Statement/i, "月度财政收支"],
];
const LOW_KEEP = /MBA Mortgage Applications|10-Year Note Auction|30-Year Bond Auction/i;
const MED_DROP = /MBA (30-Year|Mortgage Market|Mortgage Refinance|Purchase)|EIA (?!Crude Oil Stocks Change)|CPI s\.a|^CPI$/i;

function zhName(en: string): string {
  for (const [re, zh] of ZH) if (re.test(en)) return zh;
  const auction = en.match(/(\d+)-Year (?:Note|Bond) Auction/i);
  if (auction) return `${auction[1]}年期国债拍卖`;
  return en;
}

function fnum(x: unknown): string {
  if (x === null || x === undefined || x === "") return "";
  const v = Number(x);
  return Number.isFinite(v) ? `${v}` : String(x);
}

// 美东今天(YYYY-MM-DD + 周几)
function todayET(): { date: string; weekday: string } {
  const parts = new Intl.DateTimeFormat("en-CA", {
    timeZone: "America/New_York", year: "numeric", month: "2-digit", day: "2-digit", weekday: "short",
  }).formatToParts(new Date());
  const get = (t: string) => parts.find((p) => p.type === t)?.value || "";
  const wd = { Mon: "周一", Tue: "周二", Wed: "周三", Thu: "周四", Fri: "周五", Sat: "周六", Sun: "周日" }[get("weekday")] || "";
  return { date: `${get("year")}-${get("month")}-${get("day")}`, weekday: wd };
}

function etTimeOf(utc: string): { date: string; hm: string } | null {
  // Finnhub time = "YYYY-MM-DD HH:MM:SS"(UTC)
  const m = utc.match(/^(\d{4})-(\d{2})-(\d{2}) (\d{2}):(\d{2}):(\d{2})$/);
  if (!m) return null;
  const d = new Date(Date.UTC(+m[1], +m[2] - 1, +m[3], +m[4], +m[5], +m[6]));
  const parts = new Intl.DateTimeFormat("en-CA", {
    timeZone: "America/New_York", year: "numeric", month: "2-digit", day: "2-digit",
    hour: "2-digit", minute: "2-digit", hour12: false,
  }).formatToParts(d);
  const get = (t: string) => parts.find((p) => p.type === t)?.value || "";
  let hh = get("hour"); if (hh === "24") hh = "00";
  return { date: `${get("year")}-${get("month")}-${get("day")}`, hm: `${hh}:${get("minute")}` };
}

async function usNames(): Promise<Record<string, { name: string; mcapB: number | null }>> {
  try {
    const dir = path.join(process.cwd(), "public", "data");
    // 名称取 us-stocks;市值取 us-fundamentals(带 generated_at,与个股详情页 hero 同源 → 跨页市值一致)
    const [stocksRaw, fundRaw] = await Promise.all([
      fs.readFile(path.join(dir, "us-stocks.json"), "utf-8"),
      fs.readFile(path.join(dir, "us-fundamentals.json"), "utf-8").catch(() => "{}"),
    ]);
    const fund = (JSON.parse(fundRaw).stocks || {}) as Record<string, { mcapB?: number | null }>;
    const out: Record<string, { name: string; mcapB: number | null }> = {};
    for (const s of JSON.parse(stocksRaw).stocks || []) {
      out[s.sym] = { name: s.name || "", mcapB: fund[s.sym]?.mcapB ?? s.mcapB ?? null };
    }
    return out;
  } catch {
    return {};
  }
}

export async function loadTodayEvents(): Promise<TodayEvents | null> {
  const key = process.env.FINNHUB_KEY;
  if (!key) return null;
  const { date, weekday } = todayET();
  try {
    const d7 = new Date(Date.parse(date + "T12:00:00Z") + 7 * 86400_000).toISOString().slice(0, 10);
    const [ecoR, earR, ipoR, names] = await Promise.all([
      fetch(`https://finnhub.io/api/v1/calendar/economic?token=${key}`,
            { next: { revalidate: 1800 } }).then((r) => r.json()),
      fetch(`https://finnhub.io/api/v1/calendar/earnings?from=${date}&to=${date}&token=${key}`,
            { next: { revalidate: 1800 } }).then((r) => r.json()),
      fetch(`https://finnhub.io/api/v1/calendar/ipo?from=${date}&to=${d7}&token=${key}`,
            { next: { revalidate: 1800 } }).then((r) => r.json()),
      usNames(),
    ]);

    // —— 宏观(降噪 + 中文化)——
    const seen = new Set<string>();
    const macro: MacroEvent[] = [];
    for (const e of ecoR.economicCalendar || []) {
      if (e.country !== "US" && e.country !== "United States") continue;
      const t = etTimeOf(String(e.time || ""));
      if (!t || t.date !== date) continue;
      const en = String(e.event || ""), imp = String(e.impact || "");
      if (imp === "low" && !LOW_KEEP.test(en)) continue;
      if (imp === "medium" && MED_DROP.test(en)) continue;
      if (imp === "high" && /CPI s\.a|^CPI$/.test(en)) continue;
      const name = zhName(en);
      const k = `${t.hm}|${name}`;
      if (seen.has(k)) continue;
      seen.add(k);
      macro.push({ t: t.hm, hi: imp === "high", name, est: fnum(e.estimate), prev: fnum(e.prev) });
    }
    macro.sort((a, b) => a.t.localeCompare(b.t) || Number(b.hi) - Number(a.hi));

    // —— 财报(盘前/盘后)——
    const ears: { symbol?: string; hour?: string; epsEstimate?: number }[] = earR.earningsCalendar || [];
    const row = (e: (typeof ears)[number]): EarnRow => {
      const sym = e.symbol || "";
      const n = names[sym];
      return { sym, name: (n?.name || "").slice(0, 28), mcapB: n?.mcapB ?? null,
               eps: typeof e.epsEstimate === "number" ? e.epsEstimate : null };
    };
    const grp = (hours: string[], cap: number) => {
      const g = ears.filter((e) => hours.includes(e.hour || ""))
        .map(row)
        .sort((a, b) => (b.mcapB ?? 0) - (a.mcapB ?? 0));
      const main = g.filter((r) => (r.mcapB ?? 0) >= 0.5).slice(0, cap);
      const small = g.filter((r) => !main.includes(r) && r.sym).map((r) => r.sym).slice(0, 10);
      return { main, small };
    };
    const b = grp(["bmo"], 6);
    const a = grp(["amc", "", "dmh"], 7);

    // —— IPO(当天 ≥$1B;7天内 ≥$10B 巨无霸提前上)——
    const IPO_ALIAS: Record<string, string> = { SPCX: "SpaceX" };
    const WD = ["周日", "周一", "周二", "周三", "周四", "周五", "周六"];
    const ipos: IpoRow[] = [];
    for (const i of ipoR.ipoCalendar || []) {
      const valB = (i.totalSharesValue || 0) / 1e9;
      const isToday = i.date === date;
      if (!(isToday && valB >= 1) && !(valB >= 10)) continue;
      const sym = i.symbol || "";
      let nm = IPO_ALIAS[sym] || String(i.name || "").toLowerCase().replace(/\b\w/g, (c) => c.toUpperCase());
      ipos.push({ sym, name: nm.slice(0, 30), exch: String(i.exchange || "").split(" ")[0],
                  price: String(i.price || ""), valB, date: i.date, isToday });
    }
    ipos.sort((x, y) => y.valB - x.valB);

    // —— 主角 ——
    const leadBits: string[] = [];
    const hiM = macro.find((m) => m.hi);
    if (hiM) leadBits.push(`${hiM.name.split(" ")[0]} ${hiM.t}`);
    const star = [...b.main, ...a.main].sort((x, y) => (y.mcapB ?? 0) - (x.mcapB ?? 0))[0];
    if (star && (star.mcapB ?? 0) >= 20) leadBits.push(`$${star.sym} ${b.main.includes(star) ? "盘前" : "盘后"}`);
    const mega = ipos.find((i) => i.valB >= 10);
    if (mega) {
      const when = mega.isToday ? "今天" : WD[new Date(mega.date + "T12:00:00Z").getUTCDay()];
      leadBits.push(`${mega.name} IPO ${when}`);
    }

    return { date, weekday, macro, bmo: b.main, amc: a.main, smallBmo: b.small, smallAmc: a.small,
             ipos, lead: leadBits.join(" · ") };
  } catch {
    return null;
  }
}
